### PR TITLE
Create codemagic.yaml

### PR DESCRIPTION
**Pull request by ismaelAbogabal**
Basic implementation of Codemagic.yaml file 

# Implemented logic 
1. setup the machine 
2. Build an app 
3. Cache 
4. Deploy to internal testing  

# Missing login 
1. deploy into S3 bucket

### CODE REVIEW

File: codemagic.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+workflows:
+  android-tv:
+    name: Building and deploying android TV application
+    triggering:
+      events:
+        - push
+        - pull_request
```
Comment (ibrahim-zehhaf-mawaqit) [inline codemagic.yaml:7]: do not build on pull_request and push, only tag

File: codemagic.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+            --build-name=$VERSION \
+            --build-number=$BUILD_NUMBER \
+            --flavor noLeanback \
+            --dart-define mawaqit.api.key=ad283fb2-844b-40fe-967c-5cb593e9005e
+    artifacts:
+      - build/app/outputs/bundle/release/app-release.aab
+      - build/**/outputs/**/mapping.txt
+      # for the non leanback flavor build
+      - build/app/outputs/bundle/noleanbackRelease/app-noleanback-release.aab
+      - build/app/outputs/flutter-apk/app-noleanback-release.apk
```
Comment (ibrahim-zehhaf-mawaqit) [inline codemagic.yaml:52]: please rename files to be MAWAQIT-For-TV.apk .abb

File: codemagic.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        script: flutter pub get
+      - name: Build AAB with Flutter
+        script: |
+          BUILD_NUMBER=$(($(google-play get-latest-build-number --package-name "$PACKAGE_NAME" --tracks="$GOOGLE_PLAY_TRACK") + 1))
+          echo "Build Number: $BUILD_NUMBER"
+          flutter build appbundle --release \
+            --build-name=$VERSION \
+            --build-number=$BUILD_NUMBER \
+            --flavor noLeanback \
+            --dart-define mawaqit.api.key=ad283fb2-844b-40fe-967c-5cb593e9005e
```
Comment (ibrahim-zehhaf-mawaqit) [inline codemagic.yaml:38]: remove api key, should be in env var

File: codemagic.yaml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      - name: Build APK with Flutter
+        script: |
+          BUILD_NUMBER=$(($(google-play get-latest-build-number --package-name "$PACKAGE_NAME" --tracks="$GOOGLE_PLAY_TRACK") + 1))
+          flutter build apk --release \
+            --build-name=$VERSION \
+            --build-number=$BUILD_NUMBER \
+            --flavor noLeanback \
+            --dart-define mawaqit.api.key=ad283fb2-844b-40fe-967c-5cb593e9005e
```
Comment (ibrahim-zehhaf-mawaqit) [inline codemagic.yaml:46]: remove api key, should be in env var

File: codemagic.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      - name: Build APK with Flutter
+        script: |
+          BUILD_NUMBER=$(($(google-play get-latest-build-number --package-name "$PACKAGE_NAME" --tracks="$GOOGLE_PLAY_TRACK") + 1))
+          flutter build apk --release \
+            --build-name=$VERSION \
+            --build-number=$BUILD_NUMBER \
+            --flavor noLeanback \
```
Comment (ibrahim-zehhaf-mawaqit) [inline codemagic.yaml:45]: we can remove flavor we build always  with no leanback

File: codemagic.yaml (hunk 3 above)
Comment (ismaelAbogabal) [reply codemagic.yaml:38]: @ibrahim-zehhaf-mawaqit Please add it into env var and give me the key

File: codemagic.yaml (hunk 5 above)
Comment (ismaelAbogabal) [reply codemagic.yaml:45]: Ok, will remove the flavor totally because we don't need it anymore.

File: codemagic.yaml (hunk 4 above)
Comment (ibrahim-zehhaf-mawaqit) [reply codemagic.yaml:46]: in github or codemagic ?

File: codemagic.yaml (hunk 3 above)
Comment (ismaelAbogabal) [reply codemagic.yaml:38]: Codemgaic vars
MAWAQIT_API_KEY

Comment (ismaelAbogabal) [reply codemagic.yaml:38]: Codemagic key MAWAQIT_API_KEY

Comment (ibrahim-zehhaf-mawaqit) [reply codemagic.yaml:38]: done

File: codemagic.yaml (hunk 2 above)
Comment (ibrahim-zehhaf-mawaqit) [reply codemagic.yaml:52]: Or vX.X.X-buildNumber.ext
like v1.7.0-239.apk 
if possible

**Comment by ibrahim-zehhaf-mawaqit** [comment]
The version number is no more displayed 
Please add the build number too

**Comment by ismaelAbogabal** [comment]
@ibrahim-zehhaf-mawaqit

The aws_credentials isn't existing on the team variable.  

image

# References 
1. [AWS crediontials](https://docs.codemagic.io/yaml-publishing/aws/)
2. [GitHub Crediontials](https://docs.codemagic.io/yaml-publishing/github-releases/)

**Comment by ibrahim-zehhaf-mawaqit** [comment]
> @ibrahim-zehhaf-mawaqit
> 
> The aws_credentials isn't existing on the team variable.
> 
> image
> 
> # References
> 1. [AWS crediontials](https://docs.codemagic.io/yaml-publishing/aws/)
> 2. [GitHub Crediontials](https://docs.codemagic.io/yaml-publishing/github-releases/)

They are in the app settings, you need to share theme with all the team

image

**Comment by ibrahim-zehhaf-mawaqit** [comment]
These secrets are related to android-tv app only so I prefere keep them on

File: codemagic.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+      - name: Rename Build artifact
+        script: |
+          mkdir build/artifacts;
+          cp build/app/outputs/bundle/release/app-release.aab build/artifacts/MAWAQIT-For-TV-v$VERSION-$BUILD_NUMBER.aab
+          cp build/app/outputs/flutter-apk/app-release.apk build/artifacts/MAWAQIT-For-TV-v$VERSION-$BUILD_NUMBER.apk
+
+      - name: Upload to AWS S3 bucket
+        script: |
+          aws s3 build/app/outputs/flutter-apk/app-release.apk s3://cdn.mawaqit.net/android/apk/MAWAQIT-For-TV-v$VERSION-$BUILD_NUMBER.apk
+          aws s3 build/app/outputs/flutter-apk/app-release.apk s3://cdn.mawaqit.net/android/apk/MAWAQIT-For-TV-latest.apk
```
Comment (ibrahim-zehhaf-mawaqit) [inline codemagic.yaml:55]: don't push latest plz, I do it manually when I consider the release stable

File: codemagic.yaml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    environment:
+      flutter: stable
+      vars:
+        PACKAGE_NAME: "com.mawaqit.androidtv"
+        GOOGLE_PLAY_TRACK: internal
+      android_signing:
+        - Mawaqit Mobile App + Android TV Android keystore
+      groups:
+        - google_credentials
+        - aws_credentials
```
Comment (ibrahim-zehhaf-mawaqit) [inline codemagic.yaml:19]: you can add the other groups ? APP, GITHUB, AWS

**Comment by ismaelAbogabal** [comment]
@ibrahim-zehhaf-mawaqit Now all the process is working fine. 

- GitHub release 
- aws s3 deployment
- google deploy 

Please check out the s3 bucket.

File: codemagic.yaml [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+      - name: Copy artifacts
+        script: |
+          mkdir build/artifacts;
+          cp build/app/outputs/bundle/release/app-release.aab build/artifacts/MAWAQIT-For-TV-v$VERSION-$BUILD_NUMBER.aab
+          cp build/app/outputs/flutter-apk/app-release.apk build/artifacts/MAWAQIT-For-TV-v$VERSION-$BUILD_NUMBER.apk
+
+      - name: Upload to AWS S3 bucket
+        script: |
+          aws s3 cp build/artifacts/MAWAQIT-For-TV-v$VERSION-$BUILD_NUMBER.apk s3://cdn.mawaqit.net/android/apk/
```
[the remark this answers is not in the document]
Comment (ismaelAbogabal) [reply codemagic.yaml:64]: Version extractor method isn't working on mac machine 
Will find a workaround to make it work.

[the remark this answers is not in the document]
Comment (ismaelAbogabal) [reply codemagic.yaml:81]: @ibrahim-zehhaf-mawaqit Its fixed now 
image

**Pull request merged**